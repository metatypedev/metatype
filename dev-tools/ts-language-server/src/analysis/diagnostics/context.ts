// import { Diagnostic } from "vscode-languageserver";
import { Diagnostic, DiagnosticSeverity } from "vscode-languageserver/types";
import { ScopeManager } from "../typescript-semantic/scope.ts";
import { Parser } from "../../parser.ts";
import { ExposedFunction } from "../exposed_function.ts";
import { TgTypeStruct } from "../typescript-semantic/semantic-node.ts";

export class ModuleDiagnosticsContext {
  symbolRegistry: ScopeManager;
  diagnostics: Diagnostic[] = [];

  constructor(rootNode: Parser.SyntaxNode, private uri: string) {
    this.symbolRegistry = new ScopeManager(rootNode);
  }

  #pushDiagnostic(
    severity: DiagnosticSeverity,
    node: Parser.SyntaxNode,
    message: string,
  ) {
    this.diagnostics.push({
      severity,
      message,
      range: {
        start: {
          line: node.startPosition.row,
          character: node.startPosition.column,
        },
        end: {
          line: node.endPosition.row,
          character: node.endPosition.column,
        },
      },
      source: "typegraph",
    });
  }

  checkExposedFunction(exposedFunction: ExposedFunction) {
    const inputType = exposedFunction.inputType;
    if (!(inputType instanceof TgTypeStruct)) {
      this.error(
        exposedFunction.node,
        `Input type of exposed function must be a struct, but got ${inputType}`,
      );
    }
  }

  error(node: Parser.SyntaxNode, message: string) {
    this.#pushDiagnostic(DiagnosticSeverity.Error, node, message);
  }

  warn(node: Parser.SyntaxNode, message: string) {
    this.#pushDiagnostic(DiagnosticSeverity.Warning, node, message);
  }

  info(node: Parser.SyntaxNode, message: string) {
    this.#pushDiagnostic(DiagnosticSeverity.Information, node, message);
  }

  hint(node: Parser.SyntaxNode, message: string) {
    this.#pushDiagnostic(DiagnosticSeverity.Hint, node, message);
  }
}
