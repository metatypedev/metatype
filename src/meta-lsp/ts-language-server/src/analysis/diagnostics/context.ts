// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import { Diagnostic, DiagnosticSeverity } from "vscode-languageserver-types";
import { ScopeManager } from "../typescript-semantic/scope.ts";
import {
  Parser,
  TypegraphDefinition,
  TypegraphDefinitionCaptures,
} from "../../parser.ts";
import { TgTypeStruct } from "../typescript-semantic/semantic-node.ts";

export class ModuleDiagnosticsContext {
  symbolRegistry: ScopeManager;
  diagnostics: Diagnostic[] = [];

  constructor(rootNode: Parser.SyntaxNode, private uri: string) {
    this.symbolRegistry = new ScopeManager(rootNode);
  }

  #pushDiagnostic(
    severity: DiagnosticSeverity,
    node: Parser.SyntaxNode,
    message: string,
  ) {
    this.diagnostics.push({
      severity,
      message,
      range: {
        start: {
          line: node.startPosition.row,
          character: node.startPosition.column,
        },
        end: {
          line: node.endPosition.row,
          character: node.endPosition.column,
        },
      },
      source: "typegraph",
    });
  }

  public checkTypegraph(def: TypegraphDefinitionCaptures) {
    const typegraphDef = TypegraphDefinition.create(def, this);

    if (typegraphDef == null) {
      return;
    }

    for (const [name, exposedFunction] of typegraphDef.exposedFunctions) {
      const input = exposedFunction.input;
      if (input && !(input.type instanceof TgTypeStruct)) {
        this.error(
          input?.spec ?? exposedFunction.node,
          `Exposed function '${name}': expected input type to be a struct but got ${input?.type}`,
        );
      }
    }
  }

  error(node: Parser.SyntaxNode, message: string) {
    this.#pushDiagnostic(DiagnosticSeverity.Error, node, message);
  }

  warn(node: Parser.SyntaxNode, message: string) {
    this.#pushDiagnostic(DiagnosticSeverity.Warning, node, message);
  }

  info(node: Parser.SyntaxNode, message: string) {
    this.#pushDiagnostic(DiagnosticSeverity.Information, node, message);
  }

  hint(node: Parser.SyntaxNode, message: string) {
    this.#pushDiagnostic(DiagnosticSeverity.Hint, node, message);
  }
}
