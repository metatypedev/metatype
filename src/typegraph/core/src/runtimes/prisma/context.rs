// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use std::{
    cell::{OnceCell, Ref, RefCell, RefMut},
    collections::HashMap,
    rc::{Rc, Weak},
};

use crate::{typegraph::TypegraphContext, wit::runtimes as wit};
use common::typegraph::runtimes::prisma as cm;
use indexmap::{map::Entry, IndexMap, IndexSet};

use super::relationship::discovery::CandidatePair;
use super::relationship::RelationshipModel;
use super::{
    model::{InjectionHandler, Property, RelationshipProperty, ScalarProperty},
    relationship::discovery::RelationshipName,
};
use crate::errors::Result;
use crate::types::TypeId;

use super::{errors, model::Model, relationship::Relationship};

#[derive(Debug, Clone)]
pub struct ModelRef(Rc<RefCell<Model>>);

impl ModelRef {
    pub fn borrow(&self) -> Ref<Model> {
        self.0.borrow()
    }

    pub fn borrow_mut(&self) -> RefMut<Model> {
        self.0.borrow_mut()
    }

    pub fn type_name(&self) -> String {
        self.0.borrow().type_name.clone()
    }

    pub fn type_id(&self) -> TypeId {
        self.0.borrow().type_id
    }
}

impl From<Rc<RefCell<Model>>> for ModelRef {
    fn from(model: Rc<RefCell<Model>>) -> Self {
        Self(model)
    }
}

#[derive(Default, Debug)]
pub struct PrismaContext {
    models: IndexMap<TypeId, ModelRef>,
    pub models_by_name: IndexMap<String, TypeId>,
    pub relationships: IndexMap<String, Relationship>,
    pub typegen_cache: OnceCell<Weak<RefCell<HashMap<String, TypeId>>>>, // shared
    complete_registrations: IndexSet<TypeId>,
}

impl PrismaContext {
    pub fn model(&self, type_id: TypeId) -> Result<ModelRef> {
        let model = self
            .models
            .get(&type_id)
            .ok_or_else(|| errors::unregistered_model(type_id))
            .unwrap(); // TODO ?

        Ok(model.clone())
    }

    pub fn is_registered(&self, pair: &CandidatePair) -> Result<bool> {
        let left_model = pair.0.model.borrow();
        let right_model = pair.1.model.borrow();

        let left = self.models.contains_key(&left_model.type_id)
            && self.models_by_name.contains_key(&left_model.type_name)
            && left_model.relationships.contains_key(&pair.1.field_name);

        let right = self.models.contains_key(&right_model.type_id)
            && self.models_by_name.contains_key(&right_model.type_name)
            && right_model.relationships.contains_key(&pair.0.field_name);

        match (left, right) {
            (true, true) => Ok(true),
            (true, false) | (false, true) => {
                Err(format!("Pair partially registered: pair={pair:?}").into())
            }
            (false, false) => Ok(false),
        }
    }

    pub fn register_relationship(
        &mut self,
        name: RelationshipName,
        relationship: Relationship,
    ) -> Result<()> {
        use indexmap::map::Entry as E;
        use RelationshipName::*;
        match name {
            User(name) => match self.relationships.entry(name.clone()) {
                E::Occupied(rel) => {
                    let rel = rel.get();
                    return Err(format!("relationship name '{}' already used between {} and {}, please provide another name",
                        name, rel.left.model_name, rel.right.model_name).into());
                }
                E::Vacant(e) => {
                    e.insert(relationship);
                }
            },

            Generated(name) => match self.relationships.entry(name.clone()) {
                E::Occupied(rel) => {
                    let rel = rel.get();
                    return Err(format!("relationship name '{}' already used between {} and {}, please provide a name",
                                    name, rel.left.model_name, rel.right.model_name).into());
                }
                E::Vacant(e) => {
                    e.insert(relationship);
                }
            },
        }

        Ok(())
    }

    pub fn register_pair(&mut self, pair: CandidatePair) -> Result<bool> {
        if !self.is_registered(&pair)? {
            let pair = pair.ordered()?;

            let rel_name = pair.rel_name()?;
            let CandidatePair(left, right) = pair;

            let relationship = Relationship {
                name: rel_name.to_string(),
                left: RelationshipModel {
                    model_type: left.model.type_id(),
                    model_name: left.model.type_name(),
                    wrapper_type: left.property.wrapper_type_id,
                    cardinality: left.property.quantifier,
                    field: right.field_name.clone(),
                },
                right: RelationshipModel {
                    model_type: right.model.type_id(),
                    model_name: right.model.type_name(),
                    wrapper_type: right.property.wrapper_type_id,
                    cardinality: right.property.quantifier,
                    field: left.field_name.clone(),
                },
            };

            self.register_relationship(rel_name.clone(), relationship)?;

            {
                let mut left_model = left.model.borrow_mut();
                left_model
                    .relationships
                    .insert(right.field_name.clone(), rel_name.to_string());
            }

            {
                let mut right_model = right.model.borrow_mut();
                right_model
                    .relationships
                    .insert(left.field_name.clone(), rel_name.into());
            }

            Ok(true)
        } else {
            Ok(false)
        }
    }

    pub fn manage(&mut self, model_id: TypeId) -> Result<()> {
        if self.complete_registrations.contains(&model_id) {
            return Ok(());
        }

        let models = self.register_models(model_id)?;

        if models.is_empty() {
            // model already registered
            // relationship registration left
            let model = self.model(model_id)?;
            for pair in self.scan_model(model)? {
                self.register_pair(pair)?;
            }
            self.complete_registrations.insert(model_id);
        } else {
            for model_id in models.into_iter() {
                self.manage(model_id)?;
            }
        }

        Ok(())
    }

    /// register the model, and related models (recursively)
    /// returns the registered models
    /// returns empty if the given root model is already registered
    fn register_models(&mut self, root_model_id: TypeId) -> Result<Vec<TypeId>> {
        let model_id = root_model_id;
        if let Entry::Vacant(e) = self.models.entry(model_id) {
            let model: ModelRef = Rc::new(RefCell::new(model_id.try_into()?)).into();
            e.insert(model.clone());
            self.models_by_name
                .insert(model.borrow().type_name.clone(), model_id);

            let mut res = vec![model_id];

            // register related models
            {
                let model = model.borrow();
                for model_id in model.iter_related_models() {
                    res.extend(self.register_models(model_id)?);
                }
            }

            Ok(res)
        } else {
            Ok(vec![])
        }
    }

    fn convert_scalar_prop(
        &self,
        ctx: &mut TypegraphContext,
        key: &str,
        prop: &ScalarProperty,
        runtime_idx: u32,
    ) -> Result<cm::ScalarProperty> {
        Ok(cm::ScalarProperty {
            key: key.to_string(),
            cardinality: prop.quantifier.into(),
            type_idx: ctx
                .register_type(prop.wrapper_type_id.resolve_ref()?.1, Some(runtime_idx))?
                .into(),
            prop_type: prop.prop_type.clone(),
            injection: prop.injection.as_ref().map(|inj| cm::ManagedInjection {
                create: inj.create.as_ref().and_then(|handler| match handler {
                    InjectionHandler::Typegate => None,
                    InjectionHandler::PrismaDateNow => Some(cm::Injection::DateNow),
                }),
                update: inj.update.as_ref().and_then(|handler| match handler {
                    InjectionHandler::Typegate => None,
                    InjectionHandler::PrismaDateNow => Some(cm::Injection::DateNow),
                }),
            }),
            unique: prop.unique,
            auto: prop.auto,
            default_value: prop.default_value.clone(),
        })
    }

    fn convert_relationship_prop(
        &self,
        ctx: &mut TypegraphContext,
        key: &str,
        prop: &RelationshipProperty,
        rel_name: String,
        runtime_idx: u32,
    ) -> Result<cm::RelationshipProperty> {
        let model = self.model(prop.model_id)?;
        let model = model.borrow();

        Ok(cm::RelationshipProperty {
            key: key.to_string(),
            cardinality: prop.quantifier.into(),
            type_idx: ctx
                .register_type(prop.wrapper_type_id.resolve_ref()?.1, Some(runtime_idx))?
                .into(),
            model_name: model.type_name.clone(),
            unique: prop.unique,
            relationship_name: rel_name.clone(),
            relationship_side: {
                let rel = self.relationships.get(&rel_name).unwrap();
                if rel.left.wrapper_type == prop.wrapper_type_id {
                    cm::Side::Left
                } else if rel.right.wrapper_type == prop.wrapper_type_id {
                    cm::Side::Right
                } else {
                    unreachable!()
                }
            },
        })
    }

    fn convert_model(
        &self,
        ctx: &mut TypegraphContext,
        model: &Model,
        type_id: TypeId,
        runtime_idx: u32,
    ) -> Result<cm::Model> {
        Ok(cm::Model {
            type_idx: ctx
                .register_type(type_id.resolve_ref()?.1, Some(runtime_idx))?
                .into(),
            type_name: model.type_name.clone(),
            props: model
                .props
                .iter()
                .map(|(key, prop): (&String, &Property)| -> Result<_> {
                    Ok(match prop {
                        Property::Scalar(prop) => Some(cm::Property::Scalar(
                            self.convert_scalar_prop(ctx, key, prop, runtime_idx)?,
                        )),

                        Property::Model(prop) => {
                            let rel_name = {
                                let model = self.model(type_id)?;
                                let model = model.borrow();
                                model.relationships.get(key).unwrap().clone()
                            };

                            Some(cm::Property::Relationship(self.convert_relationship_prop(
                                ctx,
                                key,
                                prop,
                                rel_name,
                                runtime_idx,
                            )?))
                        }

                        Property::Unmanaged(_) => {
                            // skip
                            None
                        }
                    })
                })
                .filter_map(|r| r.transpose())
                .collect::<Result<Vec<_>>>()?,
            id_fields: model.id_fields.clone(),
            unique_constraints: model.unique_constraints.clone(),
        })
    }

    pub fn convert(
        &self,
        ctx: &mut TypegraphContext,
        runtime_idx: u32,
        data: Rc<wit::PrismaRuntimeData>,
    ) -> Result<cm::PrismaRuntimeData> {
        Ok(cm::PrismaRuntimeData {
            name: data.name.clone(),
            connection_string_secret: data.connection_string_secret.clone(),
            models: self
                .models
                .iter()
                .map(|(type_id, model)| -> Result<_> {
                    let model = model.borrow();
                    self.convert_model(ctx, &model, *type_id, runtime_idx)
                })
                .collect::<Result<Vec<_>>>()?,

            relationships: {
                self.relationships
                    .iter()
                    .map(|(_, rel)| {
                        let left =
                            self.convert_relationship_model(ctx, rel.left.clone(), runtime_idx)?;
                        let right =
                            self.convert_relationship_model(ctx, rel.right.clone(), runtime_idx)?;
                        Ok(cm::Relationship {
                            name: rel.name.clone(),
                            left,
                            right,
                        })
                    })
                    .collect::<Result<Vec<_>>>()?
            },
            migration_options: None,
        })
    }

    fn convert_relationship_model(
        &self,
        ctx: &mut TypegraphContext,
        model: RelationshipModel,
        runtime_idx: u32,
    ) -> Result<cm::RelationshipModel> {
        Ok(cm::RelationshipModel {
            type_idx: ctx
                .register_type(model.model_type.resolve_ref()?.1, Some(runtime_idx))?
                .into(),
            field: model.field.clone(),
            cardinality: model.cardinality.into(),
        })
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::test_utils::*;

    #[test]
    fn test_recursive_model_registration() -> Result<()> {
        let mut ctx = PrismaContext::default();
        let (user, profile) = models::simple_relationship()?;

        let models: Vec<_> = ctx
            .register_models(user)?
            .into_iter()
            .map(|model_id| {
                let model = ctx.model(model_id).unwrap();
                let model = model.borrow();
                (model.type_id, model.type_name.clone())
            })
            .collect();

        assert_eq!(
            models,
            vec![
                (user, user.name()?.unwrap()),
                (profile, profile.name()?.unwrap())
            ]
        );

        Ok(())
    }

    #[test]
    fn test_relationship_registration() -> Result<()> {
        let mut ctx = PrismaContext::default();
        let (user, post) = models::simple_relationship()?;

        ctx.manage(user)?;

        assert_eq!(ctx.models.len(), 2);
        let user = ctx.model(user)?;
        let user = user.borrow();
        assert_eq!(user.relationships.len(), 1);

        let post = ctx.model(post)?;
        let post = post.borrow();
        assert_eq!(post.relationships.len(), 1);

        assert!(user.relationships.get("posts").is_some());
        assert_eq!(
            user.relationships.get("posts"),
            post.relationships.get("author")
        );

        Ok(())
    }
}
