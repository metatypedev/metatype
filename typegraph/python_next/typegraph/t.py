# Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

from typegraph.gen.exports.core import (
    PolicySpecSimple,
    TypeInteger,
    TypeStruct,
    TypeFunc,
    TypeBase,
    TypeProxy,
    TypePolicy,
    PolicyPerEffect as WitPolicyPerEffect,
    PolicySpecPerEffect,
)
from typegraph.gen.types import Err
from typing import Optional, Dict, Union, Tuple
from typing_extensions import Self
from typegraph.graph.typegraph import core, store
from typegraph.runtimes.deno import Materializer
from typegraph.policy import Policy, PolicyPerEffect


class typedef:
    id: int

    def __init__(self, id: int):
        self.id = id

    def __repr__(self):
        res = core.get_type_repr(store, self.id)
        if isinstance(res, Err):
            raise Exception(res.value)
        return res.value

    def with_policy(self, *policies: Union[Policy, PolicyPerEffect]) -> Self:
        res = core.with_policy(
            store,
            TypePolicy(
                tpe=self.id,
                chain=[
                    PolicySpecSimple(value=p.id)
                    if isinstance(p, Policy)
                    else PolicySpecPerEffect(
                        value=WitPolicyPerEffect(
                            create=p.create and p.create.id,
                            update=p.update and p.update.id,
                            delete=p.delete and p.delete.id,
                            none=p.none and p.none.id,
                        )
                    )
                    for p in policies
                ],
            ),
        )

        if isinstance(res, Err):
            raise Exception(res.value)

        return _TypeWithPolicy(res.value, self, policies)


class _TypeWithPolicy(typedef):
    base: "typedef"
    policy: Tuple[Union[Policy, PolicyPerEffect], ...]

    def __init__(
        self,
        id: int,
        base: "typedef",
        policy: Tuple[Union[Policy, PolicyPerEffect], ...],
    ):
        super().__init__(id)
        self.base = base
        self.policy = policy

    def __getattr__(self, name):
        if name == "policy":
            return self.policy
        return getattr(self.base, name)


class ref(typedef):
    name: str

    def __init__(self, name: str):
        res = core.proxyb(store, TypeProxy(name=name))
        if isinstance(res, Err):
            raise Exception(res.value)
        super().__init__(res.value)
        self.name = name


class integer(typedef):
    min: Optional[int] = None
    max: Optional[int] = None

    def __init__(
        self,
        *,
        min: Optional[int] = None,
        max: Optional[int] = None,
        name: Optional[str] = None
    ):
        data = TypeInteger(min=min, max=max)
        res = core.integerb(store, data, TypeBase(name=name))
        if isinstance(res, Err):
            raise Exception(res.value)
        super().__init__(res.value)
        self.min = min
        self.max = max


class struct(typedef):
    props: Dict[str, typedef]

    def __init__(self, props: Dict[str, typedef], *, name: Optional[str] = None):
        data = TypeStruct(props=list((name, tpe.id) for (name, tpe) in props.items()))

        res = core.structb(store, data, base=TypeBase(name=name))
        if isinstance(res, Err):
            raise Exception(res.value)
        super().__init__(res.value)
        self.props = props


class func(typedef):
    inp: struct
    out: typedef
    mat: Materializer

    def __init__(self, inp: struct, out: typedef, mat: Materializer):
        data = TypeFunc(inp=inp.id, out=out.id, mat=mat.id)
        res = core.funcb(store, data)
        if isinstance(res, Err):
            raise Exception(res.value)
        id = res.value
        super().__init__(id)
        self.inp = inp
        self.out = out
