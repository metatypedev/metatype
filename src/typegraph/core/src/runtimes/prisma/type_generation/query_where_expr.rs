// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::runtimes::prisma::context::PrismaContext;
use crate::t::{self, ConcreteTypeBuilder, TypeBuilder};
use crate::{errors::Result, types::TypeId};

use super::{where_::Where, TypeGen};

pub struct QueryWhereExpr {
    model_id: TypeId,
    unique: bool,
}

impl QueryWhereExpr {
    pub fn new(model_id: TypeId) -> Self {
        Self {
            model_id,
            unique: false,
        }
    }

    pub fn unique(mut self) -> Self {
        self.unique = true;
        self
    }
}

// TODO require at least one unique field for unique queries
// Currently, this is not supported by `t.struct()`
//
// https://www.prisma.io/docs/orm/reference/prisma-client-reference#filter-on-non-unique-fields-with-userwhereuniqueinput
//
// ```
// if model.id_fields.contains(k) || prop.unique {
//   // ...
// }
// ```

impl TypeGen for QueryWhereExpr {
    fn generate(&self, context: &PrismaContext) -> Result<TypeId> {
        let where_type = context.generate(&Where::new(self.model_id))?;

        let props = where_type.as_struct().unwrap().data.props.to_vec();

        let name = self.name();
        let self_ref = t::ref_(&name).build()?;
        let and = t::optionalx(t::list(self_ref))?.build()?;

        let mut builder = t::struct_();
        builder.named(name);
        for (k, ty) in props.into_iter() {
            builder.prop(k.clone(), ty.into());
        }
        builder.prop("AND", and);
        builder.prop("OR", and);
        builder.propx("NOT", t::optional(self_ref))?;

        builder.build()
    }

    fn name(&self) -> String {
        let unique = if self.unique { "_unique" } else { "" };
        format!(
            "{}_query_where{unique}_input",
            self.model_id.name().unwrap().unwrap()
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_utils::*;
    #[test]
    fn test_query_where_expr() -> Result<()> {
        setup(None)?;

        let mut ctx = PrismaContext::default();

        let (user, post) = models::simple_relationship()?;

        ctx.manage(user)?;
        let user_where_expr = ctx.generate(&QueryWhereExpr::new(user))?;
        insta::assert_snapshot!("User/QueryWhereExpr", tree::print(user_where_expr));

        let post_where_expr = ctx.generate(&QueryWhereExpr::new(post))?;
        insta::assert_snapshot!("Post/QueryWhereExpr", tree::print(post_where_expr));

        Ok(())
    }
}
