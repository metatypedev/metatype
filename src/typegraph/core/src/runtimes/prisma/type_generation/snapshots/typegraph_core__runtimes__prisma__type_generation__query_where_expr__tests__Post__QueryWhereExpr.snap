---
source: src/typegraph/core/src/runtimes/prisma/type_generation/query_where_expr.rs
expression: "tree::print(post_where_expr)"
---
root: struct 'QueryPostWhereInput' #78
├── [AND]: optional 'optional_76' #76
│   └── item: list 'list_75' #75
│       └── item: &QueryPostWhereInput #74
├── [NOT]: optional 'optional_77' #77
│   └── item: &QueryPostWhereInput #74
├── [OR]: optional 'optional_76' #76
│   └── item: list 'list_75' #75
│       └── item: &QueryPostWhereInput #74
├── [author]: optional 'optional_72' #72
│   └── item: struct 'UserWhere__skip_8_PostWhere' #71
│       ├── [id]: optional 'optional_60' #60
│       │   └── item: optional '_integer_filter_c' #20
│       │       └── item: union #19
│       │           ├── variant_0: either '_integer_filter' #17
│       │           │   ├── variant_0: integer #9
│       │           │   ├── variant_1: struct #12
│       │           │   │   └── [equals]: integer #9
│       │           │   ├── variant_2: struct #13
│       │           │   │   └── [not]: integer #9
│       │           │   ├── variant_3: struct #14
│       │           │   │   ├── [gt]: optional 'optional_10' #10
│       │           │   │   │   └── item: integer #9
│       │           │   │   ├── [gte]: optional 'optional_10' #10
│       │           │   │   │   └── item: integer #9
│       │           │   │   ├── [lt]: optional 'optional_10' #10
│       │           │   │   │   └── item: integer #9
│       │           │   │   └── [lte]: optional 'optional_10' #10
│       │           │   │       └── item: integer #9
│       │           │   ├── variant_4: struct #15
│       │           │   │   └── [in]: list 'list_11' #11
│       │           │   │       └── item: integer #9
│       │           │   └── variant_5: struct #16
│       │           │       └── [notIn]: list 'list_11' #11
│       │           │           └── item: integer #9
│       │           └── variant_1: struct #18
│       │               └── [not]: either '_integer_filter' #17
│       │                   ├── variant_0: integer #9
│       │                   ├── variant_1: struct #12
│       │                   │   └── [equals]: integer #9
│       │                   ├── variant_2: struct #13
│       │                   │   └── [not]: integer #9
│       │                   ├── variant_3: struct #14
│       │                   │   ├── [gt]: optional 'optional_10' #10
│       │                   │   │   └── item: integer #9
│       │                   │   ├── [gte]: optional 'optional_10' #10
│       │                   │   │   └── item: integer #9
│       │                   │   ├── [lt]: optional 'optional_10' #10
│       │                   │   │   └── item: integer #9
│       │                   │   └── [lte]: optional 'optional_10' #10
│       │                   │       └── item: integer #9
│       │                   ├── variant_4: struct #15
│       │                   │   └── [in]: list 'list_11' #11
│       │                   │       └── item: integer #9
│       │                   └── variant_5: struct #16
│       │                       └── [notIn]: list 'list_11' #11
│       │                           └── item: integer #9
│       ├── [name]: optional 'optional_61' #61
│       │   └── item: optional '_string_filter_c' #37
│       │       └── item: union #36
│       │           ├── variant_0: union '_string_filter' #34
│       │           │   ├── variant_0: string #22
│       │           │   ├── variant_1: struct #25
│       │           │   │   └── [equals]: string #22
│       │           │   ├── variant_2: struct #26
│       │           │   │   └── [not]: string #22
│       │           │   ├── variant_3: struct #27
│       │           │   │   └── [in]: list 'list_24' #24
│       │           │   │       └── item: string #22
│       │           │   ├── variant_4: struct #28
│       │           │   │   └── [notIn]: list 'list_24' #24
│       │           │   │       └── item: string #22
│       │           │   ├── variant_5: struct #31
│       │           │   │   ├── [contains]: string #22
│       │           │   │   └── [mode]: optional 'optional_30' #30
│       │           │   │       └── item: string #29 enum{ '"insensitive"' }
│       │           │   ├── variant_6: struct #32
│       │           │   │   └── [search]: string #22
│       │           │   └── variant_7: struct #33
│       │           │       ├── [endsWith]: optional 'optional_23' #23
│       │           │       │   └── item: string #22
│       │           │       └── [startsWith]: optional 'optional_23' #23
│       │           │           └── item: string #22
│       │           └── variant_1: struct #35
│       │               └── [not]: union '_string_filter' #34
│       │                   ├── variant_0: string #22
│       │                   ├── variant_1: struct #25
│       │                   │   └── [equals]: string #22
│       │                   ├── variant_2: struct #26
│       │                   │   └── [not]: string #22
│       │                   ├── variant_3: struct #27
│       │                   │   └── [in]: list 'list_24' #24
│       │                   │       └── item: string #22
│       │                   ├── variant_4: struct #28
│       │                   │   └── [notIn]: list 'list_24' #24
│       │                   │       └── item: string #22
│       │                   ├── variant_5: struct #31
│       │                   │   ├── [contains]: string #22
│       │                   │   └── [mode]: optional 'optional_30' #30
│       │                   │       └── item: string #29 enum{ '"insensitive"' }
│       │                   ├── variant_6: struct #32
│       │                   │   └── [search]: string #22
│       │                   └── variant_7: struct #33
│       │                       ├── [endsWith]: optional 'optional_23' #23
│       │                       │   └── item: string #22
│       │                       └── [startsWith]: optional 'optional_23' #23
│       │                           └── item: string #22
│       └── [posts]: optional 'optional_70' #70
│           └── item: union #69
│               ├── variant_0: struct #64
│               │   └── [every]: optional 'optional_63' #63
│               │       └── item: &PostWhere #62
│               ├── variant_1: struct #66
│               │   └── [some]: optional 'optional_65' #65
│               │       └── item: &PostWhere #62
│               └── variant_2: struct #68
│                   └── [none]: optional 'optional_67' #67
│                       └── item: &PostWhere #62
├── [id]: optional 'optional_58' #58
│   └── item: optional '_integer_filter_c' #20
│       └── item: union #19
│           ├── variant_0: either '_integer_filter' #17
│           │   ├── variant_0: integer #9
│           │   ├── variant_1: struct #12
│           │   │   └── [equals]: integer #9
│           │   ├── variant_2: struct #13
│           │   │   └── [not]: integer #9
│           │   ├── variant_3: struct #14
│           │   │   ├── [gt]: optional 'optional_10' #10
│           │   │   │   └── item: integer #9
│           │   │   ├── [gte]: optional 'optional_10' #10
│           │   │   │   └── item: integer #9
│           │   │   ├── [lt]: optional 'optional_10' #10
│           │   │   │   └── item: integer #9
│           │   │   └── [lte]: optional 'optional_10' #10
│           │   │       └── item: integer #9
│           │   ├── variant_4: struct #15
│           │   │   └── [in]: list 'list_11' #11
│           │   │       └── item: integer #9
│           │   └── variant_5: struct #16
│           │       └── [notIn]: list 'list_11' #11
│           │           └── item: integer #9
│           └── variant_1: struct #18
│               └── [not]: either '_integer_filter' #17
│                   ├── variant_0: integer #9
│                   ├── variant_1: struct #12
│                   │   └── [equals]: integer #9
│                   ├── variant_2: struct #13
│                   │   └── [not]: integer #9
│                   ├── variant_3: struct #14
│                   │   ├── [gt]: optional 'optional_10' #10
│                   │   │   └── item: integer #9
│                   │   ├── [gte]: optional 'optional_10' #10
│                   │   │   └── item: integer #9
│                   │   ├── [lt]: optional 'optional_10' #10
│                   │   │   └── item: integer #9
│                   │   └── [lte]: optional 'optional_10' #10
│                   │       └── item: integer #9
│                   ├── variant_4: struct #15
│                   │   └── [in]: list 'list_11' #11
│                   │       └── item: integer #9
│                   └── variant_5: struct #16
│                       └── [notIn]: list 'list_11' #11
│                           └── item: integer #9
└── [title]: optional 'optional_59' #59
    └── item: optional '_string_filter_c' #37
        └── item: union #36
            ├── variant_0: union '_string_filter' #34
            │   ├── variant_0: string #22
            │   ├── variant_1: struct #25
            │   │   └── [equals]: string #22
            │   ├── variant_2: struct #26
            │   │   └── [not]: string #22
            │   ├── variant_3: struct #27
            │   │   └── [in]: list 'list_24' #24
            │   │       └── item: string #22
            │   ├── variant_4: struct #28
            │   │   └── [notIn]: list 'list_24' #24
            │   │       └── item: string #22
            │   ├── variant_5: struct #31
            │   │   ├── [contains]: string #22
            │   │   └── [mode]: optional 'optional_30' #30
            │   │       └── item: string #29 enum{ '"insensitive"' }
            │   ├── variant_6: struct #32
            │   │   └── [search]: string #22
            │   └── variant_7: struct #33
            │       ├── [endsWith]: optional 'optional_23' #23
            │       │   └── item: string #22
            │       └── [startsWith]: optional 'optional_23' #23
            │           └── item: string #22
            └── variant_1: struct #35
                └── [not]: union '_string_filter' #34
                    ├── variant_0: string #22
                    ├── variant_1: struct #25
                    │   └── [equals]: string #22
                    ├── variant_2: struct #26
                    │   └── [not]: string #22
                    ├── variant_3: struct #27
                    │   └── [in]: list 'list_24' #24
                    │       └── item: string #22
                    ├── variant_4: struct #28
                    │   └── [notIn]: list 'list_24' #24
                    │       └── item: string #22
                    ├── variant_5: struct #31
                    │   ├── [contains]: string #22
                    │   └── [mode]: optional 'optional_30' #30
                    │       └── item: string #29 enum{ '"insensitive"' }
                    ├── variant_6: struct #32
                    │   └── [search]: string #22
                    └── variant_7: struct #33
                        ├── [endsWith]: optional 'optional_23' #23
                        │   └── item: string #22
                        └── [startsWith]: optional 'optional_23' #23
                            └── item: string #22
